Reserve memory for host

soreana opened the pull request:
### Description

This pr is a follow up to the https://github.com/apache/cloudstack/pull/4259 quoted the following from the pr:

> By default cloudstack reserves 1Gb of RAM in hosts 
> using _dom0_memory field. Add a global setting
> "host.reserved.mem.mb" which can used to either
> increase or decrese the amount of memory which can be reserved

<!--- Describe your changes in DETAIL - And how has behaviour functionally changed. -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

<!--- ********************************************************************************* -->
<!--- NOTE: AUTOMATATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ********************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor



### How Has This Been Tested?

Please refer to the original pr at https://github.com/apache/cloudstack/pull/4259


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

soreana commented:
@blueorangutan package

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4364

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4369

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

rohityadavcloud commented:
@blueorangutan test

blueorangutan commented:
@rohityadavcloud a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-5082)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40461 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5082-kvm-centos7.zip
Smoke tests completed. 103 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 563.98 | test_kubernetes_clusters.py

DaanHoogland commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -286,6 +176,112 @@
 import com.google.common.collect.Sets;
 import com.googlecode.ipv6.IPv6Address;
 import com.googlecode.ipv6.IPv6Network;
+import org.apache.cloudstack.acl.SecurityChecker;
```
why the order of imports change? can you change that back please?

soreana replied:
I have got an OCD alert when I saw them. That is why I reorganised them :D 
+ I redo the changes.

DaanHoogland commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -472,6 +469,8 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ENABLE_DOMAIN_SETTINGS_FOR_CHILD_DOMAIN = new ConfigKey<Boolean>(Boolean.class, "enable.domain.settings.for.child.domain", "Advanced", "false",
             "Indicates whether the settings of parent domain should be applied for child domain. If true, the child domain will get value from parent domain if its not configured in child domain else global value is taken.",
             true, ConfigKey.Scope.Global, null);
+    public static final ConfigKey<Integer> HOST_RESERVED_MEM_MB = new ConfigKey<Integer>("Advanced", Integer.class, HOST_RESERVED_MEM_MB_STRING, "1024",
```
```suggestion
    public static final ConfigKey<Integer> HOST_RESERVED_MEM_MB = new ConfigKey<>("Advanced", Integer.class, HOST_RESERVED_MEM_MB_STRING, "1024",
```

soreana replied:
Good catch. Btw, the other the configs have type in angle brackets, should I remove those, as well? Maybe in separate pull request?

DaanHoogland replied:
can do in this one or a separate one. doesn´t matter.

soreana replied:
Cool, will do that in separate one. In case someone want to back port it, he/she doesn't have to get unnecessary code changes.

DaanHoogland commented on server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:
```diff
@@ -203,6 +156,52 @@
 import com.cloud.vm.dao.UserVmDetailsDao;
 import com.cloud.vm.dao.VMInstanceDao;
 import com.google.gson.Gson;
+import org.apache.cloudstack.annotation.AnnotationService;
```
can you undo the import re-order please?

soreana replied:
Done!

DaanHoogland commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -120,7 +88,43 @@
 import com.cloud.utils.nio.NioServer;
 import com.cloud.utils.nio.Task;
 import com.cloud.utils.time.InaccurateClock;
+import org.apache.cloudstack.agent.lb.IndirectAgentLB;
```
can you undo the import re-order?

soreana replied:
Done!

DaanHoogland commented:
> @DaanHoogland Can I also have this one on 4.18, please ? :)

marked for 4.18 @soreana

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6809 (SL-JID-2503)

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

soreana commented:
@blueorangutan package

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6809 (SL-JID-2504)

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4464

DaanHoogland commented:
@soreana i see an error during build:
```
[ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.18.0.0-SNAPSHOT/server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:101:8: Unused import - org.apache.cloudstack.framework.messagebus.MessageSubscriber. [UnusedImports]
```

soreana commented:
> @soreana i see an error during build:
> 
> ```
> [ERROR] /jenkins/workspace/acs-centos7-pkg-builder/dist/rpmbuild/BUILD/cloudstack-4.18.0.0-SNAPSHOT/server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:101:8: Unused import - org.apache.cloudstack.framework.messagebus.MessageSubscriber. [UnusedImports]
> ```

@DaanHoogland fixed

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6809 (SL-JID-2506)

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_check_mark: suse15. SL-JID 4466

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

soreana commented:
@blueorangutan package

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6809 (SL-JID-2508)

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/6809 (SL-JID-2509)

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 4470

DaanHoogland commented:
@blueorangutan test matrix

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins matrix job (centos7 mgmt + xs71, centos7 mgmt + vmware65, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1371,11 +1373,19 @@ public boolean configureHostParams(final Map<String, String> params) {
+            long value = Long.parseLong(params.get(HOST_RESERVED_MEM_MB_STRING));
+            s_logger.info("Reserved memory for host is " + value + "MB");
+            _dom0MinMem = value * 1024L * 1024L;
+            if (!String.valueOf(value).equals("")) {
```
```suggestion
            if (!StringUtils.isEmpty(String.valueOf(value))) {
```

blueorangutan commented:
<b>Trillian test result (tid-5134)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 39501 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5134-xenserver-71.zip
Smoke tests completed. 104 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-5135)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40850 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5135-kvm-centos7.zip
Smoke tests completed. 103 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 838.50 | test_kubernetes_clusters.py

blueorangutan commented:
<b>Trillian test result (tid-5136)</b>
Environment: vmware-65u2 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43302 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5136-vmware-65u2.zip
Smoke tests completed. 104 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@soreana I think nothing serious left. Can you answer the outstanding comments. I'll re-run the packaging/tests afterwards.

soreana commented:
@DaanHoogland Sure, I'm on it.

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 4482

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-5152)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 42880 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5152-kvm-centos7.zip
Smoke tests completed. 104 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4905

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 4980

DaanHoogland commented:
@blueorangutan test matrix

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins matrix job (centos7 mgmt + xs71, centos7 mgmt + vmware65, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-5525)</b>
Environment: xenserver-71 (x2), Advanced Networking with Mgmt server 7
Total time taken: 40347 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5525-xenserver-71.zip
Smoke tests completed. 105 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-5526)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43959 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5526-kvm-centos7.zip
Smoke tests completed. 105 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-5527)</b>
Environment: vmware-65u2 (x2), Advanced Networking with Mgmt server 7
Total time taken: 52062 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5527-vmware-65u2.zip
Smoke tests completed. 105 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please

GutoVeronezi commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -472,6 +472,8 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ENABLE_DOMAIN_SETTINGS_FOR_CHILD_DOMAIN = new ConfigKey<Boolean>(Boolean.class, "enable.domain.settings.for.child.domain", "Advanced", "false",
             "Indicates whether the settings of parent domain should be applied for child domain. If true, the child domain will get value from parent domain if its not configured in child domain else global value is taken.",
             true, ConfigKey.Scope.Global, null);
+    public static final ConfigKey<Integer> HOST_RESERVED_MEM_MB = new ConfigKey<>("Advanced", Integer.class, HOST_RESERVED_MEM_MB_STRING, "1024",
+            "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.", true);
```
As hosts can behave differently between clusters, might the operators want to set a different value for each  cluster?

GutoVeronezi replied:
cc: @wido @soreana

wido replied:
Yes, I think so. As hardware might be different and maybe you have other strategies per cluster you might want to set this differently.

weizhouapache replied:
+1.
would be good to change to a cluster setting

GutoVeronezi commented on server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:
```diff
@@ -2190,6 +2191,12 @@ protected HostVO createHostVO(final StartupCommand[] cmds, final ServerResource
+            // Update host memory reported by agent
+            if (ssCmd.getHypervisorType().equals(HypervisorType.KVM) ||
+                    ssCmd.getHypervisorType().equals(HypervisorType.LXC)) {
+                host.setDom0MinMemory(Long.parseLong(_configDao.getValue(HOST_RESERVED_MEM_MB_STRING)) * 1024L * 1024L);
```
We could use `HOST_RESERVED_MEM_MB.value()` here, instead of retrieving with `_configDao`.

soreana replied:
Done 👍

GutoVeronezi commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -450,6 +449,7 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     private Set<String> overprovisioningFactorsForValidation;
     public static final String VM_USERDATA_MAX_LENGTH_STRING = "vm.userdata.max.length";
 
+    public static final String HOST_RESERVED_MEM_MB_STRING = "host.reserved.mem.mb";
```
We could use `HOST_RESERVED_MEM_MB.key()` instead of declaring a constant with it.

soreana replied:
Done 👍

GutoVeronezi commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -121,7 +124,10 @@
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.collections.CollectionUtils;
 
+import static com.cloud.configuration.ConfigurationManagerImpl.HOST_RESERVED_MEM_MB;
+import static com.cloud.configuration.ConfigurationManagerImpl.HOST_RESERVED_MEM_MB_STRING;
```
IMHO, using the normal import would let the code more readable than using `import static`, as we would know from where the attribute is just by looking at the class qualification.

soreana replied:
Done 👍

GutoVeronezi commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1783,6 +1791,7 @@ public void processConnect(final Host host, final StartupCommand cmd, final bool
                     Map<String, String> params = new HashMap<String, String>();
                     params.put(Config.RouterAggregationCommandEachTimeout.toString(), _configDao.getValue(Config.RouterAggregationCommandEachTimeout.toString()));
                     params.put(Config.MigrateWait.toString(), _configDao.getValue(Config.MigrateWait.toString()));
+                    params.put(HOST_RESERVED_MEM_MB_STRING, _configDao.getValue(HOST_RESERVED_MEM_MB_STRING));
```
```suggestion
                    params.put(HOST_RESERVED_MEM_MB.key(), HOST_RESERVED_MEM_MB.value());
```

soreana replied:
Done 👍

GutoVeronezi commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1862,4 +1871,50 @@ public void propagateChangeToAgents(Map<String, String> params) {
+            Long zoneId = host.getDataCenterId();
+            try {
+                // Update the "ram" for all hosts
+                long updatedHostRam = (host.getTotalMemory() + host.getDom0MinMemory()) - (Integer.parseInt(value) * 1024L * 1024L);
```
We can use `ByteScaleUtils` for scale conversions.

soreana replied:
Done 👍

GutoVeronezi commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1862,4 +1871,50 @@ public void propagateChangeToAgents(Map<String, String> params) {
+                if (updatedHostRam > 0) {
+                    host.setTotalMemory(updatedHostRam);
+                    // Update "dom0_memory" in host table
+                    host.setDom0MinMemory(Integer.parseInt(value) * 1024L * 1024L);
```
We can use `ByteScaleUtils` for scale conversions.

soreana replied:
Done 👍

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1371,11 +1373,19 @@ public boolean configureHostParams(final Map<String, String> params) {
+        if (params.get(HOST_RESERVED_MEM_MB_STRING) != null) {
+            long value = Long.parseLong(params.get(HOST_RESERVED_MEM_MB_STRING));
+            s_logger.info("Reserved memory for host is " + value + "MB");
+            _dom0MinMem = value * 1024L * 1024L;
```
We can use `ByteScaleUtils` for scale conversions.

soreana replied:
Done 👍

GutoVeronezi commented on server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:
```diff
@@ -2190,6 +2191,12 @@ protected HostVO createHostVO(final StartupCommand[] cmds, final ServerResource
+            // Update host memory reported by agent
+            if (ssCmd.getHypervisorType().equals(HypervisorType.KVM) ||
+                    ssCmd.getHypervisorType().equals(HypervisorType.LXC)) {
+                host.setDom0MinMemory(Long.parseLong(_configDao.getValue(HOST_RESERVED_MEM_MB_STRING)) * 1024L * 1024L);
```
We can use `ByteScaleUtils` for scale conversions.

soreana replied:
Done 👍

wido commented:
> @nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please

Yes, sure.

However, looking at this, doesn't this make the config setting in the KVM agent.properties redundant?

soreana commented:
> > @nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please
> 
> Yes, sure.
> 
> However, looking at this, doesn't this make the config setting in the KVM agent.properties redundant?

I don't think so. The config in the agent.properties applied to one host, although this global setting applies to all hosts.

wido commented:
> > > @nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please
> > 
> > 
> > Yes, sure.
> > However, looking at this, doesn't this make the config setting in the KVM agent.properties redundant?
> 
> I don't think so. The config in the agent.properties applied to one host, although this global setting applies to all hosts.

Get it. But I don't think this can be just a global setting. You also want to be able to set this per cluster. As clusters will defer in size and configuration.

DaanHoogland commented:
@soreana can you look at the conflicts, please?

soreana commented:
> > > > @nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please
> > > 
> > > 
> > > Yes, sure.
> > > However, looking at this, doesn't this make the config setting in the KVM agent.properties redundant?
> > 
> > 
> > I don't think so. The config in the agent.properties applied to one host, although this global setting applies to all hosts.
> 
> Get it. But I don't think this can be just a global setting. You also want to be able to set this per cluster. As clusters will defer in size and configuration.

That is a good point. What would be the best? Should I make it cluster-level or add another cluster-level setting and keep the global one?

wido commented:
> > > > > @nvazquez @weizhouapache @GutoVeronezi @wido we need more reviews here, please
> > > > 
> > > > 
> > > > Yes, sure.
> > > > However, looking at this, doesn't this make the config setting in the KVM agent.properties redundant?
> > > 
> > > 
> > > I don't think so. The config in the agent.properties applied to one host, although this global setting applies to all hosts.
> > 
> > 
> > Get it. But I don't think this can be just a global setting. You also want to be able to set this per cluster. As clusters will defer in size and configuration.
> 
> @wido That is a good point. What would be the best? Should I make it cluster-level or add another cluster-level setting and keep the global one?

We can add granularity, right? So it can be global, but also per cluster, correct? If not, I'd suggest cluster only.

DaanHoogland commented:
> > @wido That is a good point. What would be the best? Should I make it cluster-level or add another cluster-level setting and keep the global one?
> 
> We can add granularity, right? So it can be global, but also per cluster, correct? If not, I'd suggest cluster only.

@wido @soreana , we don´t have true granularity, but we can, if the scope is cluster, edit at global level a vaule that will then function as a default on cluster level. No need to keep the global level to be able to set it globally. I think though, that the global level will only function as default to those clusters that have not set it yet. I am not sure about the reset.

wido commented:
> > > @wido That is a good point. What would be the best? Should I make it cluster-level or add another cluster-level setting and keep the global one?
> > 
> > 
> > We can add granularity, right? So it can be global, but also per cluster, correct? If not, I'd suggest cluster only.
> 
> @wido @soreana , we don´t have true granularity, but we can, if the scope is cluster, edit at global level a vaule that will then function as a default on cluster level. No need to keep the global level to be able to set it globally. I think though, that the global level will only function as default to those clusters that have not set it yet. I am not sure about the reset.

Yes, that seem fine to me. I think you want to do this per cluster, but a global default sounds great.

GutoVeronezi commented:
> @wido @soreana , we don´t have true granularity, but we can, if the scope is cluster, edit at global level a vaule that will then function as a default on cluster level. No need to keep the global level to be able to set it globally. I think though, that the global level will only function as default to those clusters that have not set it yet. I am not sure about the reset.

@DaanHoogland, regarding the behavior of the configurations, once we change a configuration on cluster level or another scope (except global), an entry in `cloud.<X>_details` (`cluster_details` for `Cluster`, `data_center_details` for `Zone`, and so on) is persisted, and then, for that scope, this entry will guide the behavior instead of the entry in `cloud.configurations` (global level). For resetting those configurations to the global (default) value, we can use the API [resetConfiguration](https://cloudstack.apache.org/api/apidocs-4.17/apis/resetConfiguration.html).

DaanHoogland commented:
> > @wido @soreana , we don´t have true granularity, but we can, if the scope is cluster, edit at global level a vaule that will then function as a default on cluster level. No need to keep the global level to be able to set it globally. I think though, that the global level will only function as default to those clusters that have not set it yet. I am not sure about the reset.
> 
> @DaanHoogland, regarding the behavior of the configurations, once we change a configuration on cluster level or another scope (except global), an entry in `cloud.<X>_details` (`cluster_details` for `Cluster`, `data_center_details` for `Zone`, and so on) is persisted, and then, for that scope, this entry will guide the behavior instead of the entry in `cloud.configurations` (global level). For resetting those configurations to the global (default) value, we can use the API [resetConfiguration](https://cloudstack.apache.org/api/apidocs-4.17/apis/resetConfiguration.html).

Exactly, which should be good enough. Editing the global level should set the default on a scoped level. It might be that a scoped level that has not set this setting, keeps using the default value from the global level though, instead of the edited global value. I have not tested that and if it does that would be a bug.

GutoVeronezi commented:
> Exactly, which should be good enough. Editing the global level should set the default on a scoped level. It might be that a scoped level that has not set this setting, keeps using the default value from the global level though, instead of the edited global value. I have not tested that and if it does that would be a bug.
> 
> a work around for this would be to edit the global default.

It tries to retrieve the value from the scope:

https://github.com/apache/cloudstack/blob/ca7e1ac1abbe78513bfdd57b990d58e018f8d65d/framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKey.java#L153-L164

If it returns null (meaning the value is not defined for the scope), it will retrieve the value of the global configuration:

https://github.com/apache/cloudstack/blob/ca7e1ac1abbe78513bfdd57b990d58e018f8d65d/framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKey.java#L143-L151

It will only use the default value if the global configuration is null (which should not happen) or its value is null.

DaanHoogland commented:
@soreana can you fix the conflicts and address @GutoVeronezi 's comments please?

soreana commented:
@DaanHoogland Sure, I'm working on that in this sprint.

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

DaanHoogland commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1862,4 +1871,49 @@ public void propagateChangeToAgents(Map<String, String> params) {
     }
+
+    @Override
+    public void updateCapacityOfHosts() {
```
@soreana can you simplify this method; i.e. extract some code to smaller methods.

soreana replied:
Done. :+1:

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 5309

DaanHoogland commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1346,11 +1357,19 @@ public boolean configureHostParams(final Map<String, String> params) {
             storage.persist("vm.migrate.wait", String.valueOf(intValue));
             _migrateWait = intValue;
         }
+        if (params.get(ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key()) != null) {
+            long value = Long.parseLong(params.get(ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key()));
+            s_logger.info("Reserved memory for host is " + value + "MB");
+            _dom0MinMem = ByteScaleUtils.mebibytesToBytes(value);
+            if (!StringUtils.isEmpty(String.valueOf(value))) {
+                storage.persist(ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key(), String.valueOf(value));
+            }
+        }
```
can you create a new method for this?

soreana replied:
Done 👍

DaanHoogland commented:
@soreana will you still change the new setting to be cluste scoped?

soreana commented:
> @soreana will you still change the new setting to be cluste scoped?

Yes, @DaanHoogland I'm working on it. If the schedule is tight, we can merge this one, and I will create a follow-up PR to move it to the cluster level.

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

soreana commented:
@DaanHoogland I've addressed your comments and moved the setting to cluster level.
The pr is ready for further testing.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 5317

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 5318

wido commented:
Overall it is looking good.

What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?

soreana commented:
@wido That in my view 

> Overall it is looking good.
> 
> What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?

@wido That is a good idea, but I'm not sure if there is a good way to validate the settings. Do you have any suggestion, where I can run those validations?

weizhouapache commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -990,7 +992,16 @@ public boolean configure(final String name, final Map<String, Object> params) th
-        _dom0MinMem = ByteScaleUtils.mebibytesToBytes(AgentPropertiesFileHandler.getPropertyValue(AgentProperties.HOST_RESERVED_MEM_MB));
+        long reservedMemory = AgentPropertiesFileHandler.getPropertyValue(AgentProperties.HOST_RESERVED_MEM_MB);
+
+        value = (String)params.get(ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key());
```
these lines are not needed I think.

@GutoVeronezi 
can you confirm ?

weizhouapache replied:
@soreana 
if the value is 0 or less than 0, use the default (1024) ?

weizhouapache commented:
> @wido That in my view
> 
> > Overall it is looking good.
> > What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?
> 
> @wido That is a good idea, but I'm not sure if there is a good way to validate the settings. Do you have any suggestion, where I can run those validations?

@soreana 
you can add the validation to `validateConfigurationValue` method in `ConfigurationManagerImpl` class.


a upper limit makes sense.
a lower limit might be unnecessary if there is a validation to make sure the value is greater than 0.

DaanHoogland commented:
@weizhouapache @wido @GutoVeronezi is this ready?/are you satisfied?/ did any of you test?

DaanHoogland commented:
> @wido That in my view
> 
> > Overall it is looking good.
> > What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?
> 
> @wido That is a good idea, but I'm not sure if there is a good way to validate the settings. Do you have any suggestion, where I can run those validations?

@soreana can you add a setting for that?

@wido @weizhouapache is this really needed? this is an operator action anyway, isn´t it? they should know how to tune their cloud....!?!

wido commented:
> > @wido That in my view
> > > Overall it is looking good.
> > > What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?
> > 
> > 
> > @wido That is a good idea, but I'm not sure if there is a good way to validate the settings. Do you have any suggestion, where I can run those validations?
> 
> @soreana can you add a setting for that? 😋
> 
> @wido @weizhouapache is this really needed? this is an operator action anyway, isn´t it? they should know how to tune their cloud....!?!

No, not really needed. New PR is fine.

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

vladimirpetrov commented:
Manual setting of values works fine but the reset functionality doesn't change the value in database:

```
(localcloud) 🐱 > update configuration name=host.reserved.mem.mb value=2048
{
  "configuration": {
    "category": "Advanced",
    "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
    "isdynamic": true,
    "name": "host.reserved.mem.mb",
    "value": "2048"
  }
}

(localcloud) 🐱 > list configurations name=host.reserved.mem.mb
{
  "configuration": [
    {
      "category": "Advanced",
      "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
      "isdynamic": true,
      "name": "host.reserved.mem.mb",
      "value": "2048"
    }
  ],
  "count": 1
}

mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 1073741824
        ram: 7234621440
*************************** 2. row ***************************
dom0_memory: 1073741824
        ram: 7234629632
2 rows in set (0,00 sec)
```
But when the 'reset' functionality is used:
```
(localcloud) 🐱 > reset configuration name=host.reserved.mem.mb
{
  "configuration": {
    "category": "Advanced",
    "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
    "isdynamic": true,
    "name": "host.reserved.mem.mb",
    "value": "1024"
  }
}
```
The value is changed only in the global setting, but not updated in the database:
```
(localcloud) 🐱 > list configurations name=host.reserved.mem.mb
{
  "configuration": [
    {
      "category": "Advanced",
      "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
      "isdynamic": true,
      "name": "host.reserved.mem.mb",
      "value": "1024"
    }
  ],
  "count": 1
}

mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 2147483648
        ram: 6160879616
*************************** 2. row ***************************
dom0_memory: 2147483648
        ram: 6160887808
2 rows in set (0,00 sec)
```

DaanHoogland commented:
> Manual setting of values works fine but the reset functionality doesn't change the value in database:
> 
> ```
> (localcloud) 🐱 > update configuration name=host.reserved.mem.mb value=2048
> {
>   "configuration": {
>     "category": "Advanced",
>     "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>     "isdynamic": true,
>     "name": "host.reserved.mem.mb",
>     "value": "2048"
>   }
> }
> 
> (localcloud) 🐱 > list configurations name=host.reserved.mem.mb
> {
>   "configuration": [
>     {
>       "category": "Advanced",
>       "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>       "isdynamic": true,
>       "name": "host.reserved.mem.mb",
>       "value": "2048"
>     }
>   ],
>   "count": 1
> }
> 
> mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
> *************************** 1. row ***************************
> dom0_memory: 1073741824
>         ram: 7234621440
> *************************** 2. row ***************************
> dom0_memory: 1073741824
>         ram: 7234629632
> 2 rows in set (0,00 sec)
> ```
> 
> But when the 'reset' functionality is used:
> 
> ```
> (localcloud) 🐱 > reset configuration name=host.reserved.mem.mb
> {
>   "configuration": {
>     "category": "Advanced",
>     "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>     "isdynamic": true,
>     "name": "host.reserved.mem.mb",
>     "value": "1024"
>   }
> }
> ```
> 
> The value is changed only in the global setting, but not updated in the database:
> 
> ```
> (localcloud) 🐱 > list configurations name=host.reserved.mem.mb
> {
>   "configuration": [
>     {
>       "category": "Advanced",
>       "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>       "isdynamic": true,
>       "name": "host.reserved.mem.mb",
>       "value": "1024"
>     }
>   ],
>   "count": 1
> }
> 
> mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
> *************************** 1. row ***************************
> dom0_memory: 2147483648
>         ram: 6160879616
> *************************** 2. row ***************************
> dom0_memory: 2147483648
>         ram: 6160887808
> 2 rows in set (0,00 sec)
> ```

@soreana can you look at this?

blueorangutan commented:
<b>Trillian test result (tid-5952)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43104 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5952-kvm-centos7.zip
Smoke tests completed. 105 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_create_pvlan_network | `Error` | 0.05 | test_pvlan.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 618.32 | test_vpc_vpn.py

DaanHoogland commented:
> Trillian test result (tid-5952) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 43104 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t5952-kvm-centos7.zip Smoke tests completed. 105 look OK, 2 have errors, 0 did not run Only failed and skipped tests results shown below:
> Test 	Result 	Time (s) 	Test File
> test_create_pvlan_network 	`Error` 	0.05 	test_pvlan.py
> test_01_redundant_vpc_site2site_vpn 	`Failure` 	618.32 	test_vpc_vpn.py

@soreana , and at this please?

soreana commented:
> > @wido That in my view
> > > Overall it is looking good.
> > > What I am missing is a validation that nobody can set a value <0 or maybe a super large like 1TB or example. Do we maybe need to set some lower and upper limits?
> > 
> > 
> > @wido That is a good idea, but I'm not sure if there is a good way to validate the settings. Do you have any suggestion, where I can run those validations?
> 
> @soreana can you add a setting for that? 😋
> 
> @wido @weizhouapache is this really needed? this is an operator action anyway, isn´t it? they should know how to tune their cloud....!?!

@DaanHoogland I have mixed feelings about new global settings. On one hand, I like to prevent any damage. On the other hand, I love to let admins who don't know how to tune their data center screw it. 😈 

Anyway, It would be better to merge this on 4.18 and work on the new global setting in a different milestone. I've created the #7129 to track that. Please assign it to me and plan it for the 4.18.1 milestone.


@DaanHoogland @wido, @weizhouapache, @vladimirpetrov Thanks for testing and commenting. I'm working on a fix to see what went wrong with the database value and the test cases.

DaanHoogland commented:
@soreana , will you work on this before RC cutting? (I might postpone another week, but really want to more forward)

soreana commented:
> @soreana , will you work on this before RC cutting? (I might postpone another week, but really want to more forward)

@DaanHoogland Yes, I'm working on this PR. If nothing goes wrong, I can manage to fix it 🤞 
Due to the VM reinstall issue last week, I couldn't work on this.

DaanHoogland commented:
> @DaanHoogland Yes, I'm working on this PR. If nothing goes wrong, I can manage to fix it crossed_fingers
> Due to the VM reinstall issue last week, I couldn't work on this.

So when are you ready, +-? keep in mind that we need time to test and merge as well)

soreana commented:
@blueorangutan package

blueorangutan commented:
@soreana a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-6592)<b/>

blueorangutan commented:
@DaanHoogland a [LL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-6597)<b/>

blueorangutan commented:
<b>[LL]Trillian test result (tid-6551)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 21178 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t6551-kvm-centos7.zip
Smoke tests completed. 86 look OK, 0 have errors, 24 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
all_test_safe_shutdown | `Skipped` | --- | test_safe_shutdown.py
all_test_metrics_api | `Skipped` | --- | test_metrics_api.py
all_test_outofbandmanagement | `Skipped` | --- | test_outofbandmanagement.py
all_test_outofbandmanagement_nestedplugin | `Skipped` | --- | test_outofbandmanagement_nestedplugin.py
all_test_routers_iptables_default_policy | `Skipped` | --- | test_routers_iptables_default_policy.py
all_test_secondary_storage | `Skipped` | --- | test_secondary_storage.py
all_test_service_offerings | `Skipped` | --- | test_service_offerings.py
all_test_storage_policy | `Skipped` | --- | test_storage_policy.py
all_test_templates | `Skipped` | --- | test_templates.py
all_test_update_security_group | `Skipped` | --- | test_update_security_group.py
all_test_usage_events | `Skipped` | --- | test_usage_events.py
all_test_vm_autoscaling | `Skipped` | --- | test_vm_autoscaling.py
all_test_vm_deployment_planner | `Skipped` | --- | test_vm_deployment_planner.py
all_test_vm_life_cycle | `Skipped` | --- | test_vm_life_cycle.py
all_test_vm_lifecycle_unmanage_import | `Skipped` | --- | test_vm_lifecycle_unmanage_import.py
all_test_vm_snapshot_kvm | `Skipped` | --- | test_vm_snapshot_kvm.py
all_test_vm_snapshots | `Skipped` | --- | test_vm_snapshots.py
all_test_volumes | `Skipped` | --- | test_volumes.py
all_test_vpc_ipv6 | `Skipped` | --- | test_vpc_ipv6.py
all_test_vpc_redundant | `Skipped` | --- | test_vpc_redundant.py
all_test_vpc_router_nics | `Skipped` | --- | test_vpc_router_nics.py
all_test_vpc_vpn | `Skipped` | --- | test_vpc_vpn.py
all_test_host_maintenance | `Skipped` | --- | test_host_maintenance.py
all_test_hostha_kvm | `Skipped` | --- | test_hostha_kvm.py

DaanHoogland commented:
@blueorangutan test

shwstppr commented:
@soreana can you please address merge conflicts and any outstanding comments? Should this be considered for 4.19.0.0?

soreana commented:
> @soreana can you please address merge conflicts and any outstanding comments? Should this be considered for 4.19.0.0?

@shwstppr please consider it for 4.19.0.0, I will address the merge conflict later tomorrow.

DaanHoogland commented:
@blueorangutan packge

soreana commented:
@blueorangutan packge

blueorangutan commented:
@soreana a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 7485

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-8076)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43953 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6809-t8076-kvm-centos7.zip
Smoke tests completed. 113 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

DaanHoogland commented:
@soreana I am moving this forwards to 4.20. If you see chance to work on it, feel free to readjust the milestone.

soreana commented:
> @soreana I am moving this forwards to 4.20. If you see chance to work on it, feel free to readjust the milestone.

@DaanHoogland The PR is ready for the review, and I wish we can keep it in 4.19. I would ask for review let's see if we can get enough approval to merge it in the 4.19. (The PR was here since 4.16 😄)

DaanHoogland commented:
> > @soreana I am moving this forwards to 4.20. If you see chance to work on it, feel free to readjust the milestone.
> 
> @DaanHoogland The PR is ready for the review, and I wish we can keep it in 4.19. I would ask for review let's see if we can get enough approval to merge it in the 4.19. (The PR was here since 4.16 😄)

Than why did you add the label 'needs-love'? It just needs review, right?

DaanHoogland reviewed (approved):
clgtm, but can do with unit-, integration- and maybe manual testing

soreana commented:
> > > @soreana I am moving this forwards to 4.20. If you see chance to work on it, feel free to readjust the milestone.
> > 
> > 
> > @DaanHoogland The PR is ready for the review, and I wish we can keep it in 4.19. I would ask for review let's see if we can get enough approval to merge it in the 4.19. (The PR was here since 4.16 😄)
> 
> Than why did you add the label 'needs-love'? It just needs review, right?

I thought `need-love` means the PR has been ignored for long time and need some attentions. 😄

DaanHoogland commented:
> I thought `need-love` means the PR has been ignored for long time and need some attentions. 😄

oh, I interpret it as the code needs love ;)

vladimirpetrov commented:
The situation is exactly the same as I commented on Jan 23, changing the global setting works properly (updating the database and agent.properties files) but the reset functionality (reset configuration) resets only the global setting, not the database and agent.properties files.

The intial value is 1024:
```
(localcloud) 🐱 > list configurations name=host.reserved.mem.mb
{
  "configuration": [
    {
      "category": "Advanced",
      "component": "ConfigurationManagerImpl",
      "defaultvalue": "1024",
      "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
      "displaytext": "Host reserved mem mb",
      "group": "Hypervisor",
      "isdynamic": true,
      "name": "host.reserved.mem.mb",
      "subgroup": "Hypervisor",
      "type": "Number",
      "value": "1024"
    }
  ],
  "count": 1
}
```
We change it to 2048:
```
(localcloud) 🐱 > update configuration name=host.reserved.mem.mb value=2048
{
  "configuration": {
    "category": "Advanced",
    "component": "ConfigurationManagerImpl",
    "defaultvalue": "1024",
    "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
    "displaytext": "Host reserved mem mb",
    "group": "Hypervisor",
    "isdynamic": true,
    "name": "host.reserved.mem.mb",
    "subgroup": "Hypervisor",
    "type": "Number",
    "value": "2048"
  }
```

The database is properly set:
```
mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 2147483648
        ram: 6154563584
*************************** 2. row ***************************
dom0_memory: 2147483648
        ram: 6154563584
2 rows in set (0.00 sec)
```

and the agent.properties files:
```
# cat /etc/cloudstack/agent/agent.properties | grep reserved
host.reserved.mem.mb=2048
```

Now we reset the global configuration parameter:
```
(localcloud) 🐱 > reset configuration name=host.reserved.mem.mb
{
  "configuration": {
    "category": "Advanced",
    "component": "ConfigurationManagerImpl",
    "defaultvalue": "1024",
    "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
    "displaytext": "Host reserved mem mb",
    "group": "Hypervisor",
    "isdynamic": true,
    "name": "host.reserved.mem.mb",
    "subgroup": "Hypervisor",
    "type": "Number",
    "value": "1024"
  }
}
```

It properly shows the original value (1024). The database is not updated though:
```
mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
*************************** 1. row ***************************
dom0_memory: 2147483648
        ram: 6154563584
*************************** 2. row ***************************
dom0_memory: 2147483648
        ram: 6154563584
2 rows in set (0.00 sec)
```

and the agent.properties files are not updated too:
```
# cat /etc/cloudstack/agent/agent.properties | grep reserved
host.reserved.mem.mb=2048
```

DaanHoogland commented:
@soreana can you comment on @vladimirpetrov 's findings?

shwstppr commented:
@blueorangutan package

weizhouapache commented:
> The situation is exactly the same as I commented on Jan 23, changing the global setting works properly (updating the database and agent.properties files) but the reset functionality (reset configuration) resets only the global setting, not the database and agent.properties files.
> 

good point @vladimirpetrov 
this needs some changes indeed.

the solution may be similar as
```
messageBus.subscribe(EventTypes.EVENT_CONFIGURATION_VALUE_EDIT
```

cc @soreana

soreana commented:
> The situation is exactly the same as I commented on Jan 23, changing the global setting works properly (updating the database and agent.properties files) but the reset functionality (reset configuration) resets only the global setting, not the database and agent.properties files.
> 
> The intial value is 1024:
> 
> ```
> (localcloud) 🐱 > list configurations name=host.reserved.mem.mb
> {
>   "configuration": [
>     {
>       "category": "Advanced",
>       "component": "ConfigurationManagerImpl",
>       "defaultvalue": "1024",
>       "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>       "displaytext": "Host reserved mem mb",
>       "group": "Hypervisor",
>       "isdynamic": true,
>       "name": "host.reserved.mem.mb",
>       "subgroup": "Hypervisor",
>       "type": "Number",
>       "value": "1024"
>     }
>   ],
>   "count": 1
> }
> ```
> 
> We change it to 2048:
> 
> ```
> (localcloud) 🐱 > update configuration name=host.reserved.mem.mb value=2048
> {
>   "configuration": {
>     "category": "Advanced",
>     "component": "ConfigurationManagerImpl",
>     "defaultvalue": "1024",
>     "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>     "displaytext": "Host reserved mem mb",
>     "group": "Hypervisor",
>     "isdynamic": true,
>     "name": "host.reserved.mem.mb",
>     "subgroup": "Hypervisor",
>     "type": "Number",
>     "value": "2048"
>   }
> ```
> 
> The database is properly set:
> 
> ```
> mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
> *************************** 1. row ***************************
> dom0_memory: 2147483648
>         ram: 6154563584
> *************************** 2. row ***************************
> dom0_memory: 2147483648
>         ram: 6154563584
> 2 rows in set (0.00 sec)
> ```
> 
> and the agent.properties files:
> 
> ```
> # cat /etc/cloudstack/agent/agent.properties | grep reserved
> host.reserved.mem.mb=2048
> ```
> 
> Now we reset the global configuration parameter:
> 
> ```
> (localcloud) 🐱 > reset configuration name=host.reserved.mem.mb
> {
>   "configuration": {
>     "category": "Advanced",
>     "component": "ConfigurationManagerImpl",
>     "defaultvalue": "1024",
>     "description": "Set an upper limit for memory in megabytes which will be reserved for host and not used for VM allocation.",
>     "displaytext": "Host reserved mem mb",
>     "group": "Hypervisor",
>     "isdynamic": true,
>     "name": "host.reserved.mem.mb",
>     "subgroup": "Hypervisor",
>     "type": "Number",
>     "value": "1024"
>   }
> }
> ```
> 
> It properly shows the original value (1024). The database is not updated though:
> 
> ```
> mysql> select dom0_memory,ram from host where type="Routing" and removed is null\G
> *************************** 1. row ***************************
> dom0_memory: 2147483648
>         ram: 6154563584
> *************************** 2. row ***************************
> dom0_memory: 2147483648
>         ram: 6154563584
> 2 rows in set (0.00 sec)
> ```
> 
> and the agent.properties files are not updated too:
> 
> ```
> # cat /etc/cloudstack/agent/agent.properties | grep reserved
> host.reserved.mem.mb=2048
> ```

@vladimirpetrov Thanks for testing the PR, I found the root cause, I'm working on that.

sureshanaparti commented:
Hi @soreana Can you please resolve the conflicts, and let me know if this PR can be targeted for 4.19.1. Thanks.

soreana commented:
@sureshanaparti This PR source is Leaseweb fork, I don't have access to it anymore.

@benj-n Can you please have a look?

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+        }
+
+        for (HostVO host : allHosts) {
+
```
```suggestion
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+
+            boolean updateWasSuccessFull = updateHostMemory(host);
+
+            if (! updateWasSuccessFull){
```
```suggestion
            if (!updateWasSuccessFull) {
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+            sendCommandToAgents(hostsByZone,
+                    hostId -> Collections.singletonMap(
+                            ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key(),
+                            ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.valueIn(_hostDao.findById(hostId).getClusterId()).toString()));
```
@benj-n `_hostDao.findById(hostId)` can return null for removed host, hope it's checked earlier & no NPE here.

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+
+        for (HostVO host : allHosts) {
+
+            boolean updateWasSuccessFull = updateHostMemory(host);
```
```suggestion
            boolean updateWasSuccessful= updateHostMemory(host);
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+    }
+
+    @Override
+    public void updateCapacityOfHosts() {
```
@benj-n this method implementation already exists, check above method.

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -208,6 +209,7 @@
 import com.cloud.vm.VmDetailConstants;
 import com.google.gson.Gson;
 
+
```
```suggestion
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
         }
     }
 
+    private long calculateAvailableMemoryOfHost(HostVO host){
```
```suggestion
    private long calculateAvailableMemoryOfHost(HostVO host) {
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+        return host.getTotalMemory() + host.getDom0MinMemory() - reservedMemory;
+    }
+
+    private void updateMemoriesInDb(HostVO host, long newMemoryValue){
```
```suggestion
    private void updateMemoriesInDb(HostVO host, long newMemoryValue) {
```

sureshanaparti commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -1883,25 +1894,121 @@ protected Map<Long, List<Long>> getHostsPerZone() {
+        capacityDao.update(memCap.getId(), memCap);
+    }
+
+    private boolean updateHostMemory(HostVO host){
```
```suggestion
    private boolean updateHostMemory(HostVO host) {
```

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1443,6 +1455,15 @@ private void validateLocalStorageUUID(String localStorageUUID) throws Configurat
         }
     }
 
+    private void updateDom0MinMem(PropertiesStorage storage, final Map<String, String> params){
```
```suggestion
    private void updateDom0MinMem(PropertiesStorage storage, final Map<String, String> params) {
```

sureshanaparti commented:
Hi @benj-n Please check & address the comments. Thanks.

benj-n commented:
@sureshanaparti updated with Cosmetic changes and double implementation fix.
@soreana, would you have any input on the comment above about `_hostDao.findById(hostId)` being null when host is removed?

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -1101,7 +1102,17 @@ public boolean configure(final String name, final Map<String, Object> params) th
+        value = (String)params.get(ConfigurationManagerImpl.HOST_RESERVED_MEM_MB.key());
+        long clusterReservedMemoryValue = NumbersUtil.parseInt(value, 1024);
+
+        if ( clusterReservedMemoryValue != 1024) {
```
```suggestion
        if (clusterReservedMemoryValue != 1024) {
```

sureshanaparti commented on server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:
```diff
@@ -499,6 +498,8 @@ public class ConfigurationManagerImpl extends ManagerBase implements Configurati
     public static final ConfigKey<Boolean> ENABLE_DOMAIN_SETTINGS_FOR_CHILD_DOMAIN = new ConfigKey<Boolean>(Boolean.class, "enable.domain.settings.for.child.domain", "Advanced", "false",
             "Indicates whether the settings of parent domain should be applied for child domain. If true, the child domain will get value from parent domain if its not configured in child domain else global value is taken.",
             true, ConfigKey.Scope.Global, null);
+    public static final ConfigKey<Integer> HOST_RESERVED_MEM_MB = new ConfigKey<>("Advanced", Integer.class, "host.reserved.mem.mb", "1024",
```
```suggestion
    public static final ConfigKey<Integer> HOST_RESERVED_MEM_MB = new ConfigKey<>("Advanced", Integer.class, "host.reserved.memory.mb", "1024",
```

sureshanaparti reviewed (approved):
minor comments, otherwise lgtm

sureshanaparti commented:
> @sureshanaparti updated with Cosmetic changes and double implementation fix. @soreana, would you have any input on the comment above about `_hostDao.findById(hostId)` being null when host is removed?

Hi @benj-n Is this targeted for 4.19.1, rebase & change target branch to 4.19.  Thanks.

benj-n commented:
@soreana I force pushed the rebase on 4.19 in Leaseweb:host-reserve-memory but I think as you're the owner of the PR, only you can update the base branch for this PR.

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el7 ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10155

benj-n commented:
@blueorangutan package

blueorangutan commented:
@benj-n a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el7 ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10180

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el7 ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10183

sureshanaparti commented:
@benj-n still see build errors

```
10:37:18 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project cloud-server: Compilation failure
10:37:18 [ERROR] /BUILD/cloudstack-4.19.1.0-SNAPSHOT/server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:[7806,127] error: cannot find symbol
10:37:18 [ERROR]   symbol:   variable AllowNonRFC1918CompliantIPs
10:37:18 [ERROR]   location: class ConfigurationManagerImpl
```

benj-n commented:
@blueorangutan package

blueorangutan commented:
@benj-n a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el7 ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10252

JoaoJandre commented:
@benj-n an error was caught by checkstyle. You'll need to fix that to build.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 11984

DaanHoogland commented on engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:
```diff
@@ -124,6 +131,8 @@
 import com.cloud.utils.nio.Task;
 import com.cloud.utils.time.InaccurateClock;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.ThreadContext;
```
```suggestion
```

DaanHoogland commented:
@benj-n are you still working on this?

DaanHoogland commented:
moving forward, due to inactivity

Pearl1594 commented:
@soreana @benj-n Any update on this?

soreana commented:
> @soreana @benj-n Any update on this?

@Pearl1594 I no longer have access to Leaseweb's CloudStack repository. Maybe @phsm can help us on this.
I can cherry-pick the code to my personal repository and create another PR.

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 13721

weizhouapache commented:
@soreana 
can you fix the conflicts, or close this PR ?

DaanHoogland commented:
I think @soreana has not been active for a while. closing. he can reopen when applicable. Or if anybody wants to take over they can rebase and resubmit.
